effects.cc

```diff
@@ -216,15 +200,11 @@ void Effects_manager::remove_weather_effects(
 	Actor *main_actor = gwin->get_main_actor();
 	Tile_coord apos = main_actor ? main_actor->get_tile()
 	                  : Tile_coord(-1, -1, -1);
-	Special_effect *each = effects;
-	while (each) {
-		Special_effect *next = each->next;
-		// See if we're far enough away.
-		if (each->is_weather() && (!dist ||
-		                           static_cast<Weather_effect *>(each)->out_of_range(apos, dist)))
-			remove_effect(each);
-		each = next;
-	}
+	effects.remove_if([&](const auto& ef) {
+		return ef->is_weather() &&
+	           (!dist ||
+				static_cast<Weather_effect*>(ef.get())->out_of_range(apos, dist));
```

Comment:
Hm. It is not related to your code, but I dislike this static_cast. I am wondering if there should be a virtual method for effects that returns false for non-weather effects, and checks for out-of-range for weather effects, allowing this whole lambda body to be just something like `return ef->is_out_of_range_weather(apos, dist);`.

Reply:
I'm not a big fan of pseudo-dynamic cast either.
Easy solution would be to just return -1 from `get_number` on non-weather effects (but it might carry some other logic someplace else, I haven't done any thorough analysis here) and refactor this whole interface either to the form you're mentioning or similar, e.g. I'm personally a fan of returning an empty optional in such situations, and this should be doable once C++17 is enabled.

Option number 2 - a bit more complicated, but might be worth a shot: store specific effects in separate data structures, e.g. weather, usecode light, others + one large list for storing just observer pointers to execute paint in the right order.
Simpler interface (no searching etc.) vs more funky adding/removal logic.
Whichever you prefer, I'd opt for the latter.